### PR TITLE
Pipeline typing in azure-core

### PR DESCRIPTION
**Pull request by lmazuel**
Just get `PipelineClient` sync/async with pyright happy.

The scope is to be pipeline only. Pipeline exchanges "requests" and "responses", and keep that as generics, and not use implementation as type. It's why there is generic everywhere, and typing allows us to confirm we're not biased against a particular HTTP implementation

### CODE REVIEW

File: sdk/core/azure-core/azure/core/pipeline/policies/_base.py [hunk 1]
```diff
@@ -61,11 +61,8 @@ class HTTPPolicy(abc.ABC, Generic[HTTPRequestTypeVar, HTTPResponseTypeVar]):
     :type next: ~azure.core.pipeline.policies.HTTPPolicy or ~azure.core.pipeline.transport.HttpTransport
     """
 
-    next: Union[
-        "HTTPPolicy[HTTPRequestTypeVar, HTTPResponseTypeVar]",
-        "HttpTransport[HTTPRequestTypeVar, HTTPResponseTypeVar]",
```
Comment (lmazuel) [inline sdk/core/azure-core/azure/core/pipeline/policies/_base.py:66]: This was an error introduced in my previous PR. `next` is always a policy, and can never be a transport (the transport is wrapped in a TransportRunner that implements a policy protocol)

File: sdk/core/azure-core/azure/core/pipeline/policies/_base_async.py [hunk 2]
```diff
@@ -48,14 +48,13 @@ class AsyncHTTPPolicy(abc.ABC, Generic[HTTPRequestTypeVar, AsyncHTTPResponseType
     :type next: ~azure.core.pipeline.policies.AsyncHTTPPolicy or ~azure.core.pipeline.transport.AsyncHttpTransport
     """
 
-    next: Union[
-        "AsyncHTTPPolicy[HTTPRequestTypeVar, AsyncHTTPResponseTypeVar]",
-        "AsyncHttpTransport[HTTPRequestTypeVar, AsyncHTTPResponseTypeVar]",
```
Comment (lmazuel) [inline sdk/core/azure-core/azure/core/pipeline/policies/_base_async.py:53]: This was an error introduced in my previous PR. `next` is always a policy, and can never be a transport (the transport is wrapped in a TransportRunner that implements a policy protocol)

File: sdk/core/azure-core/azure/core/pipeline/policies/_base.py (hunk 1 above)
Comment (xiangyan99) [reply sdk/core/azure-core/azure/core/pipeline/policies/_base.py:66]: `next` could be a policy or a transport (TransportRunner) https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/core/azure-core/azure/core/pipeline/_base.py#L138

Comment (xiangyan99) [reply sdk/core/azure-core/azure/core/pipeline/policies/_base.py:66]: Oh, I get your point. Yes. `next` could only be a policy. :)

File: sdk/core/azure-core/azure/core/pipeline/policies/_base.py [hunk 3]
```diff
@@ -61,11 +61,8 @@ class HTTPPolicy(abc.ABC, Generic[HTTPRequestTypeVar, HTTPResponseTypeVar]):
     :type next: ~azure.core.pipeline.policies.HTTPPolicy or ~azure.core.pipeline.transport.HttpTransport
```
Comment (xiangyan99) [inline sdk/core/azure-core/azure/core/pipeline/policies/_base.py:61]: And we also want to update this line, right?

File: sdk/core/azure-core/azure/core/pipeline/policies/_base_async.py [hunk 4]
```diff
@@ -48,14 +48,13 @@ class AsyncHTTPPolicy(abc.ABC, Generic[HTTPRequestTypeVar, AsyncHTTPResponseType
     :type next: ~azure.core.pipeline.policies.AsyncHTTPPolicy or ~azure.core.pipeline.transport.AsyncHttpTransport
```
Comment (xiangyan99) [inline sdk/core/azure-core/azure/core/pipeline/policies/_base_async.py:48]: same here

File: sdk/core/azure-core/azure/core/pipeline/policies/_base.py (hunk 3 above)
Comment (lmazuel) [reply sdk/core/azure-core/azure/core/pipeline/policies/_base.py:61]: Actually, the whole doc is wrong, `next` is not an init parameter. Removing the whole thing.

File: sdk/core/azure-core/azure/core/pipeline/_base_async.py [hunk 5]
```diff
@@ -62,11 +67,12 @@ async def send(self, request: PipelineRequest) -> PipelineResponse:
...
         """
         await _await_result(self._policy.on_request, request)
+        response: PipelineResponse[HTTPRequestType, AsyncHTTPResponseType]
         try:
-            response = await self.next.send(request)  # type: ignore
+            response = await self.next.send(request)
         except Exception:  # pylint: disable=broad-except
-            if not await _await_result(self._policy.on_exception, request):
-                raise
+            await _await_result(self._policy.on_exception, request)
```
Comment (annatisch) [inline sdk/core/azure-core/azure/core/pipeline/_base_async.py:74]: Doesn't this change the behaviour? Or did `on_exception` never return anything in the first place?

Comment (annatisch) [reply sdk/core/azure-core/azure/core/pipeline/_base_async.py:74]: I've just noticed that it seems odd that `on_exception` doesn't pass the actual exception...

Comment (xiangyan99) [reply sdk/core/azure-core/azure/core/pipeline/_base_async.py:74]: And on_exception only returns None. https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/core/azure-core/azure/core/pipeline/policies/_base.py#L118

File: sdk/core/azure-core/azure/core/_pipeline_client.py [hunk 6]
```diff
@@ -94,11 +100,18 @@ def __exit__(self, *exc_details):
...
-        transport = kwargs.get("transport")
-        policies = kwargs.get("policies")
-        per_call_policies = kwargs.get("per_call_policies", [])
-        per_retry_policies = kwargs.get("per_retry_policies", [])
+    def _build_pipeline(  # pylint: disable=no-self-use
+        self,
+        config: Configuration,
+        *,
+        transport: Optional[HttpTransport[HTTPRequestType, HTTPResponseType]] = None,
+        policies=None,
```
Comment (kashifkhan) [inline sdk/core/azure-core/azure/core/_pipeline_client.py:108]:
```suggestion
        policies: Optional[PoliciesType] = None,
```

File: sdk/core/azure-core/azure/core/_pipeline_client_async.py [hunk 7]
```diff
@@ -172,7 +184,11 @@ async def __aexit__(self, *args):
...
         await self._pipeline.__aexit__()
 
-    def _build_pipeline(self, config, **kwargs):  # pylint: disable=no-self-use
+    def _build_pipeline(
+        self, config, **kwargs
+    ) -> AsyncPipeline[
+        HTTPRequestType, AsyncHTTPResponseType
+    ]:  # pylint: disable=no-self-use
         transport = kwargs.get("transport")
         policies = kwargs.get("policies")
```
[the remark this answers is not in the document]
Comment (xiangyan99) [reply sdk/core/azure-core/azure/core/_pipeline_client_async.py]: Updated

**Comment by xiangyan99** [comment]
All the pipeline failures are false alarms.

**Comment by scbedd** [comment]
> All the pipeline failures are false alarms.

image

They were not **all** false alarms. I will send a PR that resolves the black formatting issues.

**Pull request merged**